classes/migration/upgrade/v3_4_0/I8933_EventLogLocalized.php

```diff
@@ -0,0 +1,243 @@
+ *
+ * @brief Adds a column to the event_log_settings table to store localized data such as a file name and drops setting_type column.
+ * In the event_log table allows null values for userId.
+ * Fixes the issue with duplicate event types and renames conflicting setting names
+ */
+
+namespace PKP\migration\upgrade\v3_4_0;
+
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+use PKP\install\DowngradeNotSupportedException;
+use PKP\log\event\PKPSubmissionEventLogEntry;
+use PKP\log\event\SubmissionFileEventLogEntry;
+use PKP\migration\Migration;
+
+abstract class I8933_EventLogLocalized extends Migration
+{
+    abstract protected function getContextTable(): string;
+
+    abstract protected function getContextIdColumn(): string;
+
+    /**
+     * Run the migration.
+     */
+    public function up(): void
+    {
+        Schema::table('event_log_settings', function (Blueprint $table) {
+            $table->string('locale', 14)->default('')->after('log_id');
+            $table->dropUnique('event_log_settings_unique');
+            $table->unique(['log_id', 'locale', 'setting_name'], 'event_log_settings_unique');
+            $table->string('setting_type', 6)->nullable()->change();
+        });
+
+        // Events can be triggered without a user, e.g., in schedule tasks
+        Schema::table('event_log', function (Blueprint $table) {
+            $table->dropForeign('event_log_user_id_foreign');
+            $table->dropIndex('event_log_user_id');
+            $table->bigInteger('user_id')->nullable()->change();
+            $table->foreign('user_id')->references('user_id')->on('users')->onDelete('cascade');
+            $table->index(['user_id'], 'event_log_user_id');
+            $table->boolean('is_translated')->change();
+        });
+
+        $this->fixConflictingSubmissionLogConstants();
+
+        // Rename ambiguous settings
+        $this->renameSettings();
+
+        // Localize existing submission file name entries
+        $sitePrimaryLocale = DB::table('site')->value('primary_locale');
+        DB::table('event_log_settings AS es')
+            ->join('event_log AS e', 'es.log_id', '=', 'e.log_id')
+            ->where('setting_name', 'filename')
+            ->whereIn('event_type', [
+                SubmissionFileEventLogEntry::SUBMISSION_LOG_FILE_UPLOAD,
+                SubmissionFileEventLogEntry::SUBMISSION_LOG_FILE_EDIT,
+                SubmissionFileEventLogEntry::SUBMISSION_LOG_FILE_REVISION_UPLOAD,
+            ])
+            ->orderBy('event_log_setting_id')
+            ->chunk(100, function (Collection $logChunks) use ($sitePrimaryLocale) {
+                $mapLocaleWithSettingIds = [];
+                foreach ($logChunks as $row) {
+                    // Get locale based on a submission file ID log entry
+                    $locale = $this->getContextPrimaryLocale($row, $sitePrimaryLocale);
```

Comment:
I think you can do this all in SQL for a significant performance boost. Do a `LEFT JOIN` to the appropriate table for each `assoc_type`, then join on journals (contexts) using `COALESCE() around the candidate IDs; that'll give you a single `journals` join regardless of what kind of `assoc_type` was present.

Reply:
Thinking further about performance: would we be able to turn this around to optimize it? Currently it's...
- for each file upload, edit, or revise event,
  - get its context ID's primary locale, and
  - update the setting's locale column to match.

If I understand it right, it would probably be a lot faster to go the other way around:
- For each context,
  - Get its primary locale, and
  - Update any file upload, edit, or revise event log entry settings to match.

Reply:
Thanks, that was also on my mind as a belated idea but I wasn't sure if the performance boost would be worth to re-implement it

Reply:
Just for reference, there are 3M entries in SciELO's event log table; 567k are in (SUBMISSION_LOG_FILE_UPLOAD, SUBMISSION_LOG_FILE_EDIT, SUBMISSION_LOG_FILE_REVISION_UPLOAD). So the cost of going through these individually would be very high.

Reply:
I'm getting context primary locale having only submission or submission file. Event log doesn't have `context_id` column, thus I need to iterate over all those entries. I meant to move statement that selects context locale out of the loop.

I've optimised migration where possible. The major change is setting chunking to 10000, which means that the update statement will run once for 10k entries. On my laptop 500k records are updated within 15 minutes. If it's not enough, I can try to create a wider join statement, a series of something like:
```sql
SELECT * FROM event_log AS e
JOIN event_log_settings AS es on e.log_id = es.log_id
JOIN submissions AS s ON e.assoc_id = s.submission_id
WHERE e.assoc_type = 1048585 
AND es.setting_name = 'filename'
AND s.context_id = '?'
```
Then updating log entries altogether, knowing the associated context and its primary locale. It might be faster but less prone to corrupted data.

Reply:
That sounds good, thanks -- please go ahead, and I'll do a sanity check with a large dataset once all upgrade-related merges are complete in order to make sure we don't regress on performance.